### PR TITLE
Allow representations to be masked, with possibly different masks for the components

### PR DESCRIPTION
**Pull request by mhvk**
This pull request is the first step in making `SkyCoord` able to deal with masks, by giving representations a mask. In general, I tried to make the behaviour similar to that of `Time`, which means, e.g., that the mask strictly refers to that of the data, not any associated differentials (similarly, for frames, the `mask` will not include information from any attributes, just like `Time` does not include information from `location`).

One difference is that the mask is *not* shared between the three components. This is because I *think* we want to be able to support `SkyCoord` of catalogues in which some of the sources do not have an associated distance. Most logical is for those just to have a masked distance, but unmasked longitude and latitude. But to support that, the components of `SphericalRepresentation` need to have independent masks. (I went some way in seeing if the angles still could have a shared mask, but ended up feeling this became too complex and fragile)

So, right now, the representations have a `.mask` property which just or's everything together, as well as a `get_mask` method, which allows one to select specific components.

EDIT: This is a follow up of #16844 ~Note that this builds on #16842 and #16844. The most important commit is the last one. Opening as draft for now to allow comments on the general idea first.~

- [X] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

### CODE REVIEW

**Comment by mhvk** [comment]
@taldcroft, @nstarman (for masking in general) and @StuartLittlefair, @adrn, @eerovaher (coordinate maintainers): this PR introducing masked representations is ready for review. Most important perhaps is the big picture question of whether you think my logic about not sharing masks is correct (and whether, for follow-up on frames and `SkyCoord`, the mask should not depend on attributes).

**Comment by mhvk** [comment]
Follow-up PR using this in coordinate frames and `SkyCoord` at #17016.

File: astropy/coordinates/representation/base.py
Comment (nstarman) [inline astropy/coordinates/representation/base.py:456]: I know that it's designed s.t. f one field is masked all are masked, but with early exiting the following should be equivalent speed-wise 🤞🏼

```suggestion
        return any(isinstance(getattr(self, c), Masked) for c in self.components)
```

File: astropy/coordinates/representation/base.py
Comment (nstarman) [inline astropy/coordinates/representation/base.py:388]: I think the error message needs to be updated.

**Review by nstarman — COMMENTED** [review]
Great to see more masked support!

**Comment by nstarman** [comment]
> Most important perhaps is the big picture question of whether you think my logic about not sharing masks is correct

I think so!

File: astropy/coordinates/representation/base.py
Comment (mhvk) [reply astropy/coordinates/representation/base.py:456]: Not sure: if this gets out of sync, there are arguably larger problems than whether `.masked` is still OK - e.g., any assignments with masked values will silently drop the mask (numpy problem, not something easy to catch...).

But happy to change it too...

File: astropy/coordinates/representation/base.py
Comment (mhvk) [reply astropy/coordinates/representation/base.py:388]: Will do (after waiting a bit more to check if there are further comments)

**Comment by eerovaher** [comment]
Have you thought about how masking differentials would work? If masking those is introduced separately then we might run into all sorts of problems with `apply_space_motion()`.

**Comment by eerovaher** [comment]
I'd like to focus on `apply_space_motion()` because that will force the representation and differential masks to interact, so if we can figure out how that is supposed to work then we shouldn't have too many things left to figure out. Your idea sounds reasonable enough, but I suggest you add some tests involving `apply_space_motion()` so it would be easier to discuss it in more detail. At this point seeing the tests would be valuable even if they fail because you haven't implemented masked differentials yet.

**Comment by mhvk** [comment]
In principle, `apply_space_motion` tests are easiers to add in the follow-up PR with `SkyCoord` (#17016), since here I don't yet have access to the frame/coordinate mask. But, actually, what I could do is add a test here that just looks at the mask of the coordinate `.ra, .dec` ; I could then expand on it and check the overall mask in #17016. Does that seem reasonable?

**Comment by eerovaher** [comment]
I suppose what I really want to discuss is if a user does call `apply_space_motion()` with a masked coordinate then would it be immediately obvious and how simple would it be to recover from that mistake? Would the user have to call `apply_space_motion()` again or could they just use `.unmask` after the fact? We should be able to discuss this even before the functionality is actually implemented.

**Comment by mhvk** [comment]
Again, all good points, thanks!

In general, the `Masked` class does not lose information -- the mask gets propagated independently from the values. Hence, as long as the underlying data are ~sensible, one can do `.unmasked` and get the result. Indeed, that is shown in the admittedly rather minimal docs I wrote for #17016 - https://github.com/astropy/astropy/pull/17016/files#diff-a96b4161d037c2582111db59569d89e60b76789259da9da1c4b662b96758fa44 

p.s. I may have made a mistake in separating out the two PRs... really, here, all that I hoped to argue is that having a shared mask for all elements of a representation is going to lead to problems. I think more tests are good and am happy to add any requested either here or to #17016.

**Comment by eerovaher** [comment]
> In general, the `Masked` class does not lose information -- the mask gets propagated independently from the values. Hence, as long as the underlying data are ~sensible, one can do `.unmasked` and get the result. Indeed, that is shown in the admittedly rather minimal docs I wrote for #17016 - https://github.com/astropy/astropy/pull/17016/files#diff-a96b4161d037c2582111db59569d89e60b76789259da9da1c4b662b96758fa44

That sounds good, but it should be documented in `docs/coordinates/apply_space_motion.rst` too.

> p.s. I may have made a mistake in separating out the two PRs...

`apply_space_motion()` forces the representation and differential masks to interact so they cannot be reviewed separately.

> ...really, here, all that I hoped to argue is that having a shared mask for all elements of a representation is going to lead to problems.

It makes sense to have a common mask for coordinates on the sky, but a separate mask for distance, and it also makes sense to have a common mask for the components of proper motion, but a separate mask for radial velocity.

**Comment by mhvk** [comment]
> That sounds good, but it should be documented in `docs/coordinates/apply_space_motion.rst` too.

That would open quite a can of worms, since then we should arguably discuss how masks propagate everywhere...  But what I think this does tell is that there should be a separate page with more general examples of how masks on coordinates behave, with `apply_space_motion` a logical example.

**Comment by mhvk** [comment]
> It makes sense to have a common mask for coordinates on the sky, but a separate mask for distance, and it also makes sense to have a common mask for the components of proper motion, but a separate mask for radial velocity.

I actually have a [separate branch](https://github.com/astropy/astropy/compare/main...mhvk:astropy:masked-representations-mostly-merged-masks?expand=1) in which I tried to keep the angular masks shared. In the end, I felt it was too fragile, e.g., in having to check for shared masks also in setting, resetting masks after a conversion from cartesian to spherical, etc. Also, I worried I might be too presumptuous about what a representation would look like (though I tried to give flexibility). E.g., should the masks on r and z in cylindrical coordinates also be merged? 

In the end, it is probably also clearer for the user if each element simply has its own mask. Coordinates are a bit different from `Time` in that one actually uses the elements, while the `jd1, jd2` for times are more hidden.

**Comment by mhvk** [comment]
OK, good point that one can separate the concerns: this PR indeed does not preclude later having some sharing of masks between the angular coordinates. Indeed, it may be helpful to rebase my branch that attempted to implement shared masks on this once it is merged; that will make clearer what the differences are.

**Comment by eerovaher** [comment]
Opening a few small self-contained pull requests instead of a single very large one is usually a good idea. However, in this case it seems to me that the best way to ensure we are coming up with a good API is to ensure `SkyCoord.apply_space_motion()` works well with masked instances. This makes me hesitant to review representations in isolation.

Large pull requests can be made more manageable by keeping the commit history clear and atomic, and from a quick glance at #17016 it seems like you've been doing that already, so I don't see any value in reviewing this pull request separately.

**Comment by mhvk** [comment]
@eerovaher - if it is easier for you to review the whole (i.e., #17016, which is built on this), that is definitely fine by me! I can add extra tests, etc., wherever is most appropriate.

**Comment by eerovaher** [comment]
`SkyCoord.apply_space_motion()` forces representations, differentials and time to interact with each other, so if we have working tests for `apply_space_motion()` with the masked versions of all of these then we can say with quite some confidence that everything works. Without tests for `apply_space_motion()` there is a chance we might be overlooking something.

**Comment by mhvk** [comment]
Good point. I'll add an `apply_space_motion` test to #17016

File: astropy/coordinates/representation/base.py
Comment (mhvk) [reply astropy/coordinates/representation/base.py:388]: I rebased to try to get reviews before feature freeze and fixed this one in the process.

File: astropy/coordinates/representation/base.py
Comment (nstarman) [reply astropy/coordinates/representation/base.py:458]: It's not a large priority, we're just not built around immutable data structures.

File: astropy/coordinates/representation/base.py
Comment (nstarman) [inline astropy/coordinates/representation/base.py:482]:
```suggestion
        attrs = self.components if not attrs else attrs
```

File: astropy/coordinates/representation/base.py
Comment (nstarman) [inline astropy/coordinates/representation/base.py:405]: Unrelated to this PR but maybe we should have a `_components` attribute that provides an iterator for the `_`-prefixed components.

File: astropy/coordinates/representation/base.py
Comment (nstarman) [inline astropy/coordinates/representation/base.py:981]:
```suggestion
        return self.__class__(
            *[getattr(self, component) for component in self.components],
            differentials=self.differentials | self._validate_differentials(differentials),
            copy=False,
```

File: astropy/coordinates/representation/spherical.py
Comment (nstarman) [inline astropy/coordinates/representation/spherical.py:543]: Unrelated? But good to add.

**Review by nstarman — COMMENTED** [review]
Small comments

File: astropy/coordinates/representation/base.py
Comment (mhvk) [reply astropy/coordinates/representation/base.py:981]: I'm fairly sure `ruff` will spread this over more lines...

File: astropy/coordinates/representation/base.py
Comment (mhvk) [reply astropy/coordinates/representation/base.py:482]: I find that less clear. Wish the `?=` operator had succeeded...

File: astropy/coordinates/representation/spherical.py
Comment (mhvk) [reply astropy/coordinates/representation/spherical.py:543]: It is the first commit - it is needed to ensure that the mask is kept correctly when transforming to `Radial`. Obviously, of very little use generally...

File: astropy/coordinates/representation/base.py
Comment (nstarman) [reply astropy/coordinates/representation/base.py:981]: If id doesn't then it should be `differentials=differentials`

Comment (mhvk) [reply astropy/coordinates/representation/base.py:981]: Oops... thanks!

Comment (mhvk) [reply astropy/coordinates/representation/base.py:981]: Ruff does split it over two lines, which feels less clear. I ended up going with removing the `self.differentials` on l.977 - separate out validating from merging seemed most logical, giving a clearer traceback if the validation fails.

I'll wait a bit for further comments -- here or on #17016 -- before pushing this up.

**Review by nstarman — APPROVED** [review]
Except the `differentials` repetition (https://github.com/astropy/astropy/pull/16845/files#r1797307453) which will be fixed, this LGTM.
Thanks @mhvk!

**Comment by mhvk** [comment]
> Except the `differentials` repetition (https://github.com/astropy/astropy/pull/16845/files#r1797307453) which will be fixed, this LGTM. 

I pushed the fixed version that I already had in #17016 here too, so that this can undergo final review.

**Comment by eerovaher** [comment]
Before I start reviewing this in detail I need to ask why should I review this at all if #17016 exists? Shouldn't I review just that?

**Comment by mhvk** [comment]
> Before I start reviewing this in detail I need to ask why should I review this at all if #17016 exists? Shouldn't I review just that?

You can certainly review #17016 if that works better for you - and I'm happy to "back-port" any suggestions you make there for `representations` to this PR -- which is here because for my thinking (and perhaps that of others) it really helped to separate the adjustments into ones for the base representations and the frames/`SkyCoord` that contain them.

**Comment by mhvk** [comment]
Following https://github.com/astropy/astropy/pull/17016#discussion_r1802153532, pushed an extra commit with a test showing that transformation of a spherical representation with a matrix properly preserves the angular mask.

**Comment by eerovaher** [comment]
Dividing work into smaller self-contained chunks is a good idea, but (as I've said multiple times already) `SkyCoord.apply_space_motion()` forces the representations and differentials to interact. Furthermore, `apply_space_motion()` and some frame transformations also introduce interactions with frame attributes (e.g. `obstime`). This pull request is therefore not self-contained and should not be reviewed in isolation.

I don't see why we shouldn't be able to merge #17016 before the feature freeze.

**Comment by mhvk** [comment]
OK, that's fine too!

**Pull request closed without merging**
